README.md
EasyConfig
==========

A simple library for loading configurations easily in Python, inspired by `flask.config`.


Installation
------------

Install `EasyConfig` with `pip`:

    $ pip install python-easyconfig

Install development version from `GitHub`:

    $ git clone https://github.com/RussellLuo/easyconfig.git
    $ cd easyconfig
    $ python setup.py install


Getting Started
---------------

### 1. load default configurations from a dictionary (or an object)

```python
    # from a dictionary
    from easyconfig import Config
    config = Config({'DEBUG': True})
    assert config.DEBUG == True

    # from an object
    from easyconfig import Config
    from yourapplication import default_config
    config = Config(default_config)
    assert config.DEBUG == True
```

### 2. load/update configurations from a dictionary

```python
    from easyconfig import Config
    config = Config()
    config.load({'PORT': 5000})
```

### 3. load/update configurations from an object

```python
    from easyconfig import Config
    from yourapplication import default_config
    config = Config()
    config.load(default_config)
```

### 4. load configurations from environment variables

```python
    import os
    os.environ.setdefault('SECRET_KEY', '123***456')

    from easyconfig import Config
    config = Config(datasrc=os.environ)
    assert config.SECRET_KEY == '123***456'
```

### 5. load/update configurations from a string pointing to an object

```python
    from easyconfig import Config, str_object
    config = Config()
    config.load(str_object('yourapplication.default_config'))
    assert config.SECRET_KEY == '123***456'
```

### 6. load/update configurations from an environment variable pointing to an object

```python
    import os
    os.environ.setdefault('EASY_CONFIG', 'yourapplication.default_config')

    from easyconfig import Config, str_object
    config = Config()
    config.load(str_object('EASY_CONFIG', is_envvar=True))
    assert config.SECRET_KEY == '123***456'
```

### 7. load/update configurations from a JSON file

```python
    from easyconfig import Config, json_mapping
    config = Config()
    config.load(json_mapping('/data/configs/default_config.json'))
    assert config.SECRET_KEY == '123***456'
```

### 8. load/update configurations from an environment variable pointing to a JSON file

```python
    import os
    os.environ.setdefault('EASY_CONFIG', '/data/configs/default_config.json')

    from easyconfig import Config, json_mapping
    config = Config()
    config.load(json_mapping('EASY_CONFIG', is_envvar=True))
    assert config.SECRET_KEY == '123***456'
```

### 9. load/update configurations from a YAML file

```python
    from easyconfig import Config, yaml_mapping
    config = Config()
    config.load(yaml_mapping('/data/configs/default_config.yml'))
    assert config.SECRET_KEY == '123***456'
```

### 10. load/update configurations from an environment variable pointing to a YAML file

```python
    import os
    os.environ.setdefault('EASY_CONFIG', '/data/configs/default_config.yml')

    from easyconfig import Config, yaml_mapping
    config = Config()
    config.load(yaml_mapping('EASY_CONFIG', is_envvar=True))
    assert config.SECRET_KEY == '123***456'
```


CHANGELOG.md
# EasyConfig Changelog

Here you can see the full list of changes between each EasyConfig release.


## Version 0.1.7

Released on Nov 2nd 2017.

- Remove the fixed version of dependencies (i.e. six and PyYAML)


## Version 0.1.6

Released on Feb 20th 2016.

- Add a `load` method to the `Config` class
- Add two keyword arguments to `str_object`
- Add `is_envvar` keyword argument to `json_mapping` and `yaml_mapping`
- Update tests to be pytest-style


## Version 0.1.5

Released on Sep 20th 2015.

- Add gitignore and makefile
- Add support for Python 2/3 compatibility
- Use built-in `import_string` instead of Werkzeug's version
- Remove the dependency on Werkzeug
- Update tests


## Version 0.1.0

Released on Jan 18th 2015.

The first release.


tests/test_utils.py
# -*- coding: utf-8 -*-

import pytest

from easyconfig.utils import import_string


class TestUtils(object):

    def test_import_a_config_module(self):
        config_module = import_string('configs.default_config')
        assert config_module.__name__ == 'configs.default_config'
        assert config_module.DEBUG is True
        assert config_module.PORT == 5000
        assert config_module.SECRET_KEY == '123***456'

    def test_import_a_nonexistent_config_module(self):
        with pytest.raises(ImportError) as exc:
            import_string('configs.dummy_config_module')
        assert str(exc.value) == (
            'No module named "configs.dummy_config_module", nor does a '
            'module named "configs" define a "dummy_config_module" attribute'
        )

    def test_import_a_config_class(self):
        config_class = import_string('configs.default_config_in_class.Config')
        assert config_class.__name__ == 'Config'
        assert config_class.DEBUG is True
        assert config_class.PORT == 5000
        assert config_class.SECRET_KEY == '123***456'

    def test_import_a_nonexistent_config_class(self):
        with pytest.raises(ImportError) as exc:
            import_string('configs.default_config_in_class.dummy_class')
        assert str(exc.value) == (
            'No module named "configs.default_config_in_class.dummy_class", '
            'nor does a module named "configs.default_config_in_class" '
            'define a "dummy_class" attribute'
        )


tests/test_config.py
# -*- coding: utf-8 -*-

import os

import pytest

from easyconfig import Config


class AppConfig(object):

    SECRET_KEY = '123***456'


class TestConfig(object):

    def test_cant_set_lowercase_item(self):
        config = Config()
        with pytest.raises(ValueError):
            config['debug'] = True

    def test_can_set_uppercase_item(self):
        config = Config()
        config['DEBUG'] = True
        assert config['DEBUG']

    def test_can_set_lowercase_attribute_as_attribute(self):
        config = Config()
        config.debug = True
        # seems like an attribute
        assert hasattr(config, 'debug')
        # actually is an attribute
        assert 'debug' in config.__dict__

    def test_cant_set_uppercase_attribute_as_attribute(self):
        config = Config()
        config.DEBUG = True
        # seems like an attribute
        assert hasattr(config, 'DEBUG')
        # actually is not an attribute
        assert 'DEBUG' not in config.__dict__

    def test_uppercase_items_are_also_attributes(self):
        config = Config()
        config.DEBUG = True
        assert config['DEBUG']

        config['PORT'] = 5000
        assert config.PORT == 5000

    def test_cant_load_lowercase_keys(self):
        config = Config({'debug': True})
        assert 'debug' not in config
        assert not hasattr(config, 'debug')

    def test_can_load_defaults_from_mapping(self):
        config = Config({'DEBUG': True})
        assert config.DEBUG

    def test_can_load_defaults_from_object(self):
        config = Config(AppConfig)
        assert config.SECRET_KEY == '123***456'

    def test_load_from_mapping(self):
        config = Config()
        config.from_mapping({'PORT': 5000})
        assert config.PORT == 5000

    def test_load_from_object(self):
        config = Config()
        config.from_object(AppConfig)
        assert config.SECRET_KEY == '123***456'

    def test_load_from_none_object(self):
        config = Config()
        config.from_object(None)
        uppercase_attrs = [attr for attr in dir(config) if attr.isupper()]
        assert not uppercase_attrs

    def test_load_from_envvar(self):
        config = Config(datasrc=os.environ)
        assert 'SECRET_KEY' not in config

        os.environ.setdefault('SECRET_KEY', '123***456')
        assert config.SECRET_KEY == '123***456'

        os.environ['SECRET_KEY'] = 'xxx***yyy'
        assert config.SECRET_KEY == 'xxx***yyy'

    def test_load_from_empty_envvar(self):
        config = Config(datasrc=os.environ)
        assert not hasattr(config, 'NULL_KEY')

    def test_load_from_either_mapping_or_object(self):
        config1 = Config()
        config1.load({'DEBUG': True})
        assert config1.DEBUG

        config2 = Config()
        config2.load(AppConfig)
        assert config2.SECRET_KEY == '123***456'


tests/test_objects.py
# -*- coding: utf-8 -*-

import os

import pytest

from easyconfig import str_object, envvar_object


class TestObjects(object):

    def test_valid_str_object(self):
        obj = str_object('configs.default_config')
        assert obj.DEBUG
        assert obj.PORT == 5000
        assert obj.SECRET_KEY == '123***456'

    def test_invalid_str_object(self):
        with pytest.raises(ValueError) as exc:
            str_object('')
        assert str(exc.value) == 'Empty module name'

    def test_invalid_str_object_in_silent(self):
        obj = str_object('', silent=True)
        assert obj is None

    def test_valid_str_object_from_envvar(self):
        os.environ['EASY_CONFIG'] = 'configs.default_config'
        obj = str_object('EASY_CONFIG', is_envvar=True)
        assert obj.DEBUG
        assert obj.PORT == 5000
        assert obj.SECRET_KEY == '123***456'

    def test_nonexistent_str_object_from_envvar(self):
        os.environ.pop('EASY_CONFIG', None)
        with pytest.raises(RuntimeError) as exc:
            str_object('EASY_CONFIG', silent=False, is_envvar=True)
        assert str(exc.value) == (
            "The environment variable 'EASY_CONFIG' is not set "
            "and as such configuration could not be "
            "loaded. Set this variable and make it "
            "point to a configuration file"
        )

    def test_nonexistent_str_object_from_envvar_in_silent(self):
        os.environ.pop('EASY_CONFIG', None)
        obj = str_object('EASY_CONFIG', silent=True, is_envvar=True)
        assert obj is None

    def test_empty_str_object_from_envvar(self):
        os.environ['EASY_CONFIG'] = ''
        with pytest.raises(ValueError) as exc:
            str_object('EASY_CONFIG', silent=False, is_envvar=True)
        assert str(exc.value) == 'Empty module name'

    def test_empty_str_object_from_envvar_in_silent(self):
        os.environ['EASY_CONFIG'] = ''
        obj = str_object('EASY_CONFIG', silent=True, is_envvar=True)
        assert obj is None

    def test_valid_envvar_object(self):
        os.environ['EASY_CONFIG'] = 'configs.default_config'
        obj = envvar_object('EASY_CONFIG')
        assert obj.DEBUG
        assert obj.PORT == 5000
        assert obj.SECRET_KEY == '123***456'


easyconfig/utils.py
import sys
from importlib import import_module

from six import reraise


def import_string(import_path):
    """Import a module path and return the module/attribute designated by
    the last name in the path. Raise ImportError if the import failed.
    """
    # The destination object is a module
    try:
        module = import_module(import_path)
    except ImportError:
        if '.' not in import_path:
            raise
    else:
        return module

    # The destination object is an attribute
    module_path, attr_name = import_path.rsplit('.', 1)
    module = import_module(module_path)
    try:
        return getattr(module, attr_name)
    except AttributeError:
        msg = (
            'No module named "{0}.{1}", nor does a module '
            'named "{0}" define a "{1}" attribute'.format(
                module_path, attr_name
            )
        )
        reraise(ImportError, ImportError(msg), sys.exc_info()[2])


easyconfig/mappings.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import json

import yaml


def load_from_file(load_func, filename, silent=False):
    """Get a configuration mapping from a file."""
    try:
        with open(filename) as fp:
            obj = load_func(fp)
            if not isinstance(obj, dict):
                raise RuntimeError(
                    'The configuration file %r is invalid' % filename
                )
    except IOError as e:
        if not silent:
            e.strerror = 'Unable to load configuration file (%s)' % e.strerror
            raise
        else:
            obj = {}

    return obj


def json_mapping(filename, silent=False):
    """Get a configuration mapping from a JSON file."""
    load_func = lambda fp: json.load(fp)
    return load_from_file(load_func, filename, silent)


def yaml_mapping(filename, silent=False):
    """Get a configuration mapping from a YAML file."""
    load_func = lambda fp: yaml.load(fp)
    return load_from_file(load_func, filename, silent)


easyconfig/config.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

from six import iteritems


class Config(dict):
    """The core configuration class.

    Example usage:

        # load configurations from a dict (as default)
        config = Config({'DEBUG': True})
        assert config['DEBUG'] == True
        assert config.DEBUG == True

        # load/update configurations from a dict
        config.from_mapping({'PORT': 5000})

        # load/update configurations from an object
        from yourapplication import default_config
        config.from_object(default_config)

        # load configurations from environment variables
        import os
        config = Config(datasrc=os.environ)
        assert 'DEBUG' not in config
        os.environ.setdefault('SECRET_KEY', '123***456')
        assert config.SECRET_KEY == '123***456'

    Note: only uppercase keys will be added as config.
    """

    def __init__(self, defaults=None, datasrc=None):
        self._datasrc = datasrc or {}
        self.from_mapping(defaults or {})

    def __getitem__(self, key):
        """Override to take `self._datasrc` into account."""
        if key in self:
            return dict.__getitem__(self, key)
        else:
            return object.__getattribute__(self, '_datasrc')[key]

    def __setitem__(self, key, value):
        """Override to only allow uppercase `key` to be set."""
        if key.isupper():
            dict.__setitem__(self, key, value)
        else:
            raise ValueError('Only uppercase keys are allowed')

    def __getattr__(self, name):
        """Override to treat the string `name` first as a dict key,
        then as an attribute name if `KeyError` is raised.
        """
        try:
            return self.__getitem__(name)
        except KeyError:
            return object.__getattribute__(self, name)

    def __setattr__(self, name, value):
        """Override to treat the pair `(name, value)` first as a dict item,
        then as an attribute if `ValueError` is raised.
        """
        try:
            self.__setitem__(name, value)
        except ValueError:
            object.__setattr__(self, name, value)

    def from_mapping(self, mapping):
        for key, value in iteritems(mapping):
            if key.isupper():
                self[key] = value

    def from_object(self, obj):
        for key in dir(obj):
            if key.isupper():
                self[key] = getattr(obj, key)
